### PR TITLE
Enable NHWC batchnorm for miopen

### PR DESCRIPTION
**Pull request by dnikolaev-amd**
* Non-static PYTORCH_MIOPEN_SUGGEST_NHWC env variable enables MIOpen batchnorm for NHWC and can be changed at runtime
* New unit test `test_batchnorm_nhwc_miopen` compares NHWC MIOpen batchnorm with native NHWC batchnorm to verify miopen. It works only with fp32 input, fp16 and fp64 will be done soon in another PR

### CODE REVIEW

**Comment by dnikolaev-amd** [comment]
New test addad. It compares NHWC MIOpen batchnorm with NHWC CPU batchnorm
PYTORCH_MIOPEN_SUGGEST_NHWC=1 python test/test_nn.py -v  TestNN.test_batchnorm_miopen_nhwc_compare_cpu

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -521,7 +522,8 @@ BatchNormBackend _select_batch_norm_backend(
         || (!running_mean.defined() && !running_var.defined() && training))
       && detail::getCUDAHooks().compiledWithMIOpen()
       && cudnn_enabled
-      && input.suggest_memory_format() != MemoryFormat::ChannelsLast
+      && ((input.suggest_memory_format() == MemoryFormat::Contiguous )
+        || (miopen_conv_use_channels_last(input, weight)))
```
Comment (jithunnair-amd) [inline aten/src/ATen/native/Normalization.cpp:526]: Looks slightly odd to use a `conv_` function in normalization code.

**Review by jithunnair-amd — COMMENTED** [review]
@jeffdaily Would it be better (esp. from upstreamability perspective) to refactor the use of the `miopen_conv_use_channels_last` function to something like `miopen_use_channels_last` for the common logic, and have any conv/batchnorm-specific logic in the respective functions?

**Comment by dnikolaev-amd** [comment]
test rewritten to use native nhwc batchnorm to verify miopen
non static PYTORCH_MIOPEN_SUGGEST_NHWC variable to be able to change it in runtime 
@jithunnair-amd @jeffdaily

**Review by jithunnair-amd — APPROVED** [review]
LGTM. Please add a snippet of the unit test output with `AMD_LOG_LEVEL=3` to show the MIOpen kernels vs native kernels being invoked.
Please also add output from a small reproducer (include reproducer code in comment) to show the env var can be turned on and off at runtime.

**Review by jithunnair-amd — CHANGES_REQUESTED** [review]
Please set the `PYTORCH_MIOPEN_SUGGEST_NHWC=1` env variable in `test_nn.py` globally, so that *all* test_nn unit tests run with this env variable and catch any issues in the MIOpen path. Please attach output from `TestNN.test_batchnorm_cudnn_nhwc` (without env var explicity set at commandline) to show that it is invoking the MIOpen path:
`CI=1 PYTORCH_TEST_WITH_ROCM=1 python test/run_test.py -v -i test_nn -k test_batchnorm_cudnn_nhwc`

**Comment by dnikolaev-amd** [comment]
We can enable and disable nhwc batchnorm on MIOpen in runtime
```python
# python -u batchnorm_nhwc_enable_miopen_in_runtime.py

import os
import torch

PYTORCH_MIOPEN_SUGGEST_NHWC = "PYTORCH_MIOPEN_SUGGEST_NHWC"

# enable MIOpen logging
os.environ["MIOPEN_ENABLE_LOGGING_CMD"] = "1"
# os.environ["AMD_LOG_LEVEL"] = "3"

print("START")
m = torch.nn.BatchNorm2d(100).to("cuda:0")

# random NCHW tensor on GPU
input = torch.randn(20, 100, 5, 4).to("cuda:0")

# change tensor dims to NHWC
input = input.to(memory_format=torch.channels_last)
print(f"input.shape={input.shape}")
print(f"input.stride={input.stride()}")

print("Disable MIOpen")
if PYTORCH_MIOPEN_SUGGEST_NHWC in os.environ:
    os.environ.clear(PYTORCH_MIOPEN_SUGGEST_NHWC)
print(f"PYTORCH_MIOPEN_SUGGEST_NHWC={os.getenv(PYTORCH_MIOPEN_SUGGEST_NHWC, default=None)}. No 'MIOpen(HIP):' output expected")
# call batch norm
output = m(input)
print()

print("Enable MIOpen")
os.environ[PYTORCH_MIOPEN_SUGGEST_NHWC] = "1"
print(f"PYTORCH_MIOPEN_SUGGEST_NHWC={os.getenv(PYTORCH_MIOPEN_SUGGEST_NHWC, default=None)}. Expected 'MIOpen(HIP):' output")
# call batch norm
output = m(input)
print()

print("Disable MIOpen")
os.environ[PYTORCH_MIOPEN_SUGGEST_NHWC] = "0"
print(f"PYTORCH_MIOPEN_SUGGEST_NHWC={os.getenv(PYTORCH_MIOPEN_SUGGEST_NHWC, default=None)}. No 'MIOpen(HIP):' output expected")
# call batch norm
output = m(input)
print()

print(f"output.shape={output.shape}")
print(f"output.stride={output.stride()}")
print("END")
``` 
output
```
input.shape=torch.Size([20, 100, 5, 4])
input.stride=(2000, 1, 400, 100)
START
Disable MIOpen
PYTORCH_MIOPEN_SUGGEST_NHWC=None. No 'MIOpen(HIP):' output expected

Enable MIOpen
PYTORCH_MIOPEN_SUGGEST_NHWC=1. Expected 'MIOpen(HIP):' output
MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 20 -c 100 -H 5 -W 4 -M 1 --forw 1 -b 0 -s 1 -r 1

Disable MIOpen
PYTORCH_MIOPEN_SUGGEST_NHWC=0. No 'MIOpen(HIP):' output expected

output.shape=torch.Size([20, 100, 5, 4])
output.stride=(2000, 1, 400, 100)
END
```

**Comment by dnikolaev-amd** [comment]
test_nn.py enables to use MIOpen for NHWC batchnorms for whole file by defining `PYTORCH_MIOPEN_SUGGEST_NHWC=1` on startup. 
`MIOPEN_ENABLE_LOGGING_CMD=1` enables miopen logging. String `MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm ...` can be observed without definition of `PYTORCH_MIOPEN_SUGGEST_NHWC=1` from command line 
```
/var/lib/jenkins/pytorch# CI=1 PYTORCH_TEST_WITH_ROCM=1 MIOPEN_ENABLE_LOGGING_CMD=1 python test/run_test.py -v -i test_nn -k test_batchnorm_nhwc_cuda 

/var/lib/jenkins/pytorch/test/run_test.py:21: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  import pkg_resources
::warning:: Gathered no stats from artifacts for build env linux-focal-rocm6.0-py3.8 build env and None test config. Using default build env and default test config instead.
::warning:: Gathered no stats from artifacts for build env linux-focal-rocm6.0-py3.8 build env and None test config. Using default build env and default test config instead.
Running all tests
Running parallel tests on 2 processes
Name: tests to run (est. time: 3.55min)
  Serial tests (1):
    test_nn 1/1
  Parallel tests (0):
Name: excluded (est. time: 0.0min)
  Serial tests (0):
  Parallel tests (0):
Starting test batch 'tests to run' 0.0 seconds after initiating testing
No TD results found
Downloading https://ossci-metrics.s3.amazonaws.com/slow-tests.json to /var/lib/jenkins/pytorch/test/.pytorch-slow-tests.json
Downloading https://ossci-metrics.s3.amazonaws.com/disabled-tests-condensed.json to /var/lib/jenkins/pytorch/test/.pytorch-disabled-tests.json
Running test_nn 1/1 ... [2024-06-04 15:54:43.585762]
Executing ['/opt/conda/envs/py_3.9/bin/python', '-bb', 'test_nn.py', '--shard-id=1', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-k', 'test_batchnorm_nhwc_cuda', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests'] ... [2024-06-04 15:54:43.586660]

test_nn 1/1 was successful, full logs can be found in artifacts with path test/test-reports/test_nn_1.1_ac755ba8f1aea1a9_.log
Running 1 items in this shard: test/test_nn.py::TestNN::test_batchnorm_nhwc_cuda

 START of temporary dumping of test_nn 1/1 execution log file from (/var/lib/jenkins/pytorch/test/test-reports/test_nn_1.1_mr9hiz5z_toprint.log)
Test results will be stored in test-reports/python-pytest/test_nn/test_nn-1c4656d3af11c97e.xml
============================= test session starts ==============================
platform linux -- Python 3.9.19, pytest-7.3.2, pluggy-1.5.0 -- /opt/conda/envs/py_3.9/bin/python
cachedir: .pytest_cache
hypothesis profile 'pytorch_ci' -> database=None, deadline=None, max_examples=50, derandomize=True, suppress_health_check=[HealthCheck.too_slow]
rootdir: /var/lib/jenkins/pytorch
configfile: pytest.ini
plugins: flakefinder-1.1.0, cpp-2.3.0, xdist-3.3.1, hypothesis-5.35.1, xdoctest-1.1.0, rerunfailures-14.0
collecting ... collected 2193 items / 2192 deselected / 1 selected
stepcurrent: Cannot find last run test, not skipping
Running 1 items in this shard: test/test_nn.py::TestNN::test_batchnorm_nhwc_cuda

test_nn.py::TestNN::test_batchnorm_nhwc_cuda MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 2 -c 64 -H 50 -W 50 -M 1 --forw 2 -b 0 -s 1
MIOpen(HIP): Command [LogCmdBNorm] ./bin/MIOpenDriver bnorm -n 2 -c 64 -H 50 -W 50 -M 1 --forw 2 -b 0 -s 1
PASSED [1.4000s]            [100%]

- generated xml file: /var/lib/jenkins/pytorch/test/test-reports/python-pytest/test_nn/test_nn-1c4656d3af11c97e.xml -
====================== 1 passed, 2192 deselected in 1.59s ======================
END of temporary dumping of test_nn 1/1 execution log file form (/var/lib/jenkins/pytorch/test/test-reports/test_nn_1.1_mr9hiz5z_toprint.log)
```

**Review by jithunnair-amd — APPROVED** [review]
